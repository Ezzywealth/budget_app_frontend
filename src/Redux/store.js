import { configureStore } from '@reduxjs/toolkit';
import budgetReducer from './BudgetSlice';
import userReducer from './UserSlice';

const store = configureStore({
	reducer: {
		budget: budgetReducer,
		user: userReducer,
	},
});

export default store;
