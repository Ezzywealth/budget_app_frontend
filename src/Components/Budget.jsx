import { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate, useParams } from 'react-router-dom';
import { fetchSingleBudget } from '../Redux/BudgetSlice';
import EditForm from './EditForm';
import { RotatingLines } from 'react-loader-spinner';

const Budget = () => {
	const navigate = useNavigate();
	const budget = useSelector((state) => state.budget.selectedBudget);
	const user = useSelector((state) => state.user.user);
	const budgetLoading = useSelector((state) => state.budget.budgetLoading);
	const [startEdit, setStartEdit] = useState(false);
	const { budgetId } = useParams();
	const dispatch = useDispatch();

	// an effect to check if the user is logged in
	useEffect(() => {
		if (!user) {
			navigate('/login');
		}
	}, [navigate, user]);

	// useEffect to fetch a single budget details
	useEffect(() => {
		dispatch(fetchSingleBudget(budgetId));
	}, [budgetId, dispatch]);

	// a function to open up the edit form
	const editBudget = () => {
		setStartEdit(true);
	};

	// a function to delete the budget

	return (
		<section className='w-full py-4 flex items-center h-screen'>
			{startEdit && <EditForm budget={budget} setStartEdit={setStartEdit} />}
			<div className='bg-gray-500 mx-4  flex flex-col w-full justify-center py-6 px-4 lg:px-8'>
				<div className='mt-8 mx-auto w-full'>
					<div className='bg-white py-8 px-2 shadow sm:rounded-lg w-[360px] md:w-[500px]'>
						<h2 className='text-3xl text-center font-extrabold text-gray-900'>Budget Details</h2>
						{budgetLoading ? (
							<div className='flex justify-center h-[500px]'>
								<RotatingLines strokeColor='#00d690' strokeWidth='5' animationDuration='0.75' width='60' visible={true} />
							</div>
						) : (
							<div className='mt-6'>
								<div className='flex justify-between items-center sm:gap-4 sm:px-6 sm:py-3'>
									<dt className='text-lg text-gray-500 font-semibold '>Name</dt>
									<dd className='mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2'>{budget?.name}</dd>
								</div>
								<div className='flex justify-between items-center sm:gap-4 sm:px-6 sm:py-3'>
									<dt className='text-lg text-gray-500 font-semibold '>Amount</dt>
									<dd className='mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2'>${budget?.amount?.toFixed(2)}</dd>
								</div>
								<div className='flex justify-between items-center sm:gap-4 sm:px-6 sm:py-3'>
									<dt className='text-lg text-gray-500 font-semibold '>Status</dt>
									<dd className='mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2'>{budget?.status}</dd>
								</div>
								<div className='flex justify-between items-center sm:gap-4 sm:px-6 sm:py-3'>
									<dt className='text-lg text-gray-500 font-semibold '>Category</dt>
									<dd className='mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2'>{budget?.category}</dd>
								</div>
								<div className='flex justify-between items-center sm:gap-4 sm:px-6 sm:py-3'>
									<dt className='text-lg text-gray-500 font-semibold '>Allow Overspending</dt>
									<dd className='mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2'>{budget?.allow_overspending ? 'Yes' : 'No'}</dd>
								</div>
								<div className='flex justify-between items-center sm:gap-4 sm:px-6 sm:py-3'>
									<dt className='text-lg text-gray-500 font-semibold '>Start Date</dt>
									<dd className='mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2'>{budget?.start_date}</dd>
								</div>
								<div className='flex justify-between items-center sm:gap-4 sm:px-6 sm:py-3'>
									<dt className='text-lg text-gray-500 font-semibold '>End Date</dt>
									<dd className='mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2'>{budget?.end_date}</dd>
								</div>
								<div className='flex justify-between items-center sm:gap-4 sm:px-6 sm:py-3'>
									<dt className='text-lg text-gray-500 font-semibold '>Note</dt>
									<dd className='mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2'>{budget?.notes}</dd>
								</div>
							</div>
						)}
					</div>
					<div className=' flex flex-col md:flex-row gap-4 px-4 my-4 md:gap-20 justify-between mt-4'>
						<button className='w-full text-white bg-[#a3915a] font-medium rounded-lg flex justify-center focus:outline-none items-center px-4 py-2 tracking-wider text-xl text-center transition-all duration-300 ease-linear' onClick={editBudget}>
							<span> Edit Budget</span>
						</button>
					</div>
				</div>
			</div>
		</section>
	);
};

export default Budget;
