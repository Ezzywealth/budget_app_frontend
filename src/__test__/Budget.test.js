import { render } from '@testing-library/react';
import Budget from '../Components/Budget'; // Adjust the import path as needed
import '@testing-library/jest-dom';
import { screen } from '@testing-library/dom';
import { describe, test, expect } from '@jest/globals';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import store from '../Redux/store';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import budgets from '../utils/data';

const mock = new MockAdapter(axios);
const mockStore = configureMockStore([thunk]);

/**
 * @jest-environment jsdom
 */

describe('Budget', () => {
	mock.onGet('http://127.0.0.1:3001/api/v1/budgetlists').reply(200, budgets[0]);
	test('renders Budget component correctly', () => {
		const { asFragment } = render(
			<Provider store={mockStore(store.getState())}>
				<BrowserRouter>
					<Budget />
				</BrowserRouter>
			</Provider>
		);
		expect(asFragment()).toMatchSnapshot();
	});

	test('renders component with Provider and BrowserRouter', () => {
		render(
			<Provider store={mockStore(store.getState())}>
				<BrowserRouter>
					<Budget />
				</BrowserRouter>
			</Provider>
		);
		// Check if the "Budget Details" heading is present

		const headingElement = screen.getByText(/Budget Details/i);
		expect(headingElement).toBeInTheDocument();

		// Check if the loading spinner is present when budgetLoading is true

		const spendingText = screen.getByText(/Allow Overspending/i); // Assuming you have a test ID for the spinner
		expect(spendingText).toBeInTheDocument();

		const categoryText = screen.getByText(/Category/i); // Assuming you have a test ID for the spinner
		expect(categoryText).toBeInTheDocument();

		const nameText = screen.getByText(/Name/i); // Assuming you have a test ID for the spinner
		expect(nameText).toBeInTheDocument();

		const dateText = screen.getByText(/End Date/i);
		expect(dateText).toBeInTheDocument();

		// Check if the "Edit Budget" button is present

		const editButton = screen.getByText(/Edit Budget/i);
		expect(editButton).toBeInTheDocument();
	});
});
