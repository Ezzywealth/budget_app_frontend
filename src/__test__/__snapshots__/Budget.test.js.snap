// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Budget renders Budget component correctly 1`] = `
<DocumentFragment>
  <section
    class="w-full py-4 flex items-center h-screen"
  >
    <div
      class="bg-gray-500 mx-4  flex flex-col w-full justify-center py-6 px-4 lg:px-8"
    >
      <div
        class="mt-8 mx-auto w-full"
      >
        <div
          class="bg-white py-8 px-2 shadow sm:rounded-lg w-[360px] md:w-[500px]"
        >
          <h2
            class="text-3xl text-center font-extrabold text-gray-900"
          >
            Budget Details
          </h2>
          <div
            class="mt-6"
          >
            <div
              class="flex justify-between items-center sm:gap-4 sm:px-6 sm:py-3"
            >
              <dt
                class="text-lg text-gray-500 font-semibold "
              >
                Name
              </dt>
              <dd
                class="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
              />
            </div>
            <div
              class="flex justify-between items-center sm:gap-4 sm:px-6 sm:py-3"
            >
              <dt
                class="text-lg text-gray-500 font-semibold "
              >
                Amount
              </dt>
              <dd
                class="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
              >
                $
              </dd>
            </div>
            <div
              class="flex justify-between items-center sm:gap-4 sm:px-6 sm:py-3"
            >
              <dt
                class="text-lg text-gray-500 font-semibold "
              >
                Status
              </dt>
              <dd
                class="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
              />
            </div>
            <div
              class="flex justify-between items-center sm:gap-4 sm:px-6 sm:py-3"
            >
              <dt
                class="text-lg text-gray-500 font-semibold "
              >
                Category
              </dt>
              <dd
                class="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
              />
            </div>
            <div
              class="flex justify-between items-center sm:gap-4 sm:px-6 sm:py-3"
            >
              <dt
                class="text-lg text-gray-500 font-semibold "
              >
                Allow Overspending
              </dt>
              <dd
                class="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
              >
                No
              </dd>
            </div>
            <div
              class="flex justify-between items-center sm:gap-4 sm:px-6 sm:py-3"
            >
              <dt
                class="text-lg text-gray-500 font-semibold "
              >
                Start Date
              </dt>
              <dd
                class="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
              />
            </div>
            <div
              class="flex justify-between items-center sm:gap-4 sm:px-6 sm:py-3"
            >
              <dt
                class="text-lg text-gray-500 font-semibold "
              >
                End Date
              </dt>
              <dd
                class="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
              />
            </div>
            <div
              class="flex justify-between items-center sm:gap-4 sm:px-6 sm:py-3"
            >
              <dt
                class="text-lg text-gray-500 font-semibold "
              >
                Note
              </dt>
              <dd
                class="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
              />
            </div>
          </div>
        </div>
        <div
          class=" flex flex-col md:flex-row gap-4 px-4 my-4 md:gap-20 justify-between mt-4"
        >
          <button
            class="w-full text-white bg-[#a3915a] font-medium rounded-lg flex justify-center focus:outline-none items-center px-4 py-2 tracking-wider text-xl text-center transition-all duration-300 ease-linear"
          >
            <span>
               Edit Budget
            </span>
          </button>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
